lib/xnvme_be_macos_driverkit_async.c

```diff
@@ -0,0 +1,219 @@
+// SPDX-FileCopyrightText: Samsung Electronics Co., Ltd
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <libxnvme.h>
+#include <xnvme_be.h>
+#include <xnvme_be_nosys.h>
+#ifdef XNVME_BE_MACOS_ENABLED
+#include <errno.h>
+#include <xnvme_be_macos_driverkit.h>
+#include <mach/mach_error.h>
+
+#include <xnvme_dev.h>
+#include <xnvme_queue.h>
+
+struct xnvme_queue_macos_driverkit {
+	struct xnvme_queue_base base;
+	RingQueue *sq_queue;
+	RingQueue *cq_queue;
+
+	uint id;
+
+	uint8_t _rsvd[208];
+};
+XNVME_STATIC_ASSERT(sizeof(struct xnvme_queue_macos_driverkit) == XNVME_BE_QUEUE_STATE_NBYTES,
+		    "Incorrect size")
+
+int
+xnvme_be_macos_driverkit_queue_init(struct xnvme_queue *_queue, int XNVME_UNUSED(opts))
+{
+	struct xnvme_queue_macos_driverkit *q = (struct xnvme_queue_macos_driverkit *)_queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)_queue->base.dev->be.state;
+	mach_vm_size_t size = 0;
+	mach_vm_address_t addr;
+	kern_return_t res;
+	int qpid;
+
+	qpid = _xnvme_be_driverkit_create_ioqpair(state, q->base.capacity, 0);
+	if (qpid < 0) {
+		XNVME_DEBUG("FAILED: nvme_create_ioqpair(); err: %d", qpid);
+		return qpid;
+	}
+	q->id = qpid;
+
+	res = IOConnectMapMemory64(state->device_connection, 1 << 16 | qpid, mach_task_self(),
+				   &addr, &size, kIOMapAnywhere);
+	XNVME_DEBUG("INFO: IOConnectMapMemory64(); returned: %x, %s", res, mach_error_string(res));
+	if (res != kIOReturnSuccess) {
+		XNVME_DEBUG("FAILED: IOConnectMapMemory64(); "
+			    "ret(0x%08x), '%s'",
+			    res, mach_error_string(res));
+		return -EIO;
+	}
+	q->sq_queue = (RingQueue *)addr;
+
+	size = 0;
+	res = IOConnectMapMemory64(state->device_connection, qpid, mach_task_self(), &addr, &size,
+				   kIOMapAnywhere);
+	XNVME_DEBUG("INFO: IOConnectMapMemory64(); returned: %x, %s", res, mach_error_string(res));
+	if (res != kIOReturnSuccess) {
+		XNVME_DEBUG("FAILED: IOConnectMapMemory64(); "
+			    "ret(0x%08x), '%s'",
+			    res, mach_error_string(res));
+		return -EIO;
+	}
+	q->cq_queue = (RingQueue *)addr;
+
+	return 0;
+}
+
+int
+xnvme_be_macos_driverkit_queue_term(struct xnvme_queue *_queue)
+{
+	struct xnvme_queue_macos_driverkit *q = (struct xnvme_queue_macos_driverkit *)_queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)q->base.dev->be.state;
+	int ret;
+
+	ret = _xnvme_be_driverkit_delete_ioqpair(state, q->id);
+	if (ret) {
+		XNVME_DEBUG("FAILED: _xnvme_be_driverkit_delete_ioqpair()");
+		return ret;
```

Comment:
Should be ``return -EIO;``, since the value in ``ret`` is not applicable for ``errno``,  right?

Reply:
No, because `_xnvme_be_driverkit_delete_ioqpair` is returning `-EIO` or `0`.

Reply:
ok, great.